File: src/com/github/yuukis/businessmap/utils/StringJUtils.java
/*
 * StringJUtils.java
 * 
 * Copyright 2013 Yuuki Shimizu.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.yuukis.businessmap.utils;

public class StringJUtils {

	private static final char[] HANKAKU_KATAKANA = { '｡', '｢', '｣', '､', '･',
			'ｦ', 'ｧ', 'ｨ', 'ｩ', 'ｪ', 'ｫ', 'ｬ', 'ｭ', 'ｮ', 'ｯ', 'ｰ', 'ｱ', 'ｲ',
			'ｳ', 'ｴ', 'ｵ', 'ｶ', 'ｷ', 'ｸ', 'ｹ', 'ｺ', 'ｻ', 'ｼ', 'ｽ', 'ｾ', 'ｿ',
			'ﾀ', 'ﾁ', 'ﾂ', 'ﾃ', 'ﾄ', 'ﾅ', 'ﾆ', 'ﾇ', 'ﾈ', 'ﾉ', 'ﾊ', 'ﾋ', 'ﾌ',
			'ﾍ', 'ﾎ', 'ﾏ', 'ﾐ', 'ﾑ', 'ﾒ', 'ﾓ', 'ﾔ', 'ﾕ', 'ﾖ', 'ﾗ', 'ﾘ', 'ﾙ',
			'ﾚ', 'ﾛ', 'ﾜ', 'ﾝ', 'ﾞ', 'ﾟ' };

	private static final char[] ZENKAKU_KATAKANA = { '。', '「', '」', '、', '・',
			'ヲ', 'ァ', 'ィ', 'ゥ', 'ェ', 'ォ', 'ャ', 'ュ', 'ョ', 'ッ', 'ー', 'ア', 'イ',
			'ウ', 'エ', 'オ', 'カ', 'キ', 'ク', 'ケ', 'コ', 'サ', 'シ', 'ス', 'セ', 'ソ',
			'タ', 'チ', 'ツ', 'テ', 'ト', 'ナ', 'ニ', 'ヌ', 'ネ', 'ノ', 'ハ', 'ヒ', 'フ',
			'ヘ', 'ホ', 'マ', 'ミ', 'ム', 'メ', 'モ', 'ヤ', 'ユ', 'ヨ', 'ラ', 'リ', 'ル',
			'レ', 'ロ', 'ワ', 'ン', '゛', '゜' };

	private static final char HANKAKU_KATAKANA_FIRST_CHAR = HANKAKU_KATAKANA[0];

	private static final char HANKAKU_KATAKANA_LAST_CHAR = HANKAKU_KATAKANA[HANKAKU_KATAKANA.length - 1];

	/**
	 * 半角カタカナから全角カタカナへ変換します。
	 *
	 * @param c
	 *            変換前の文字
	 * @return 変換後の文字
	 */
	public static char hankakuKatakanaToZenkakuKatakana(char c) {
		if (c >= HANKAKU_KATAKANA_FIRST_CHAR && c <= HANKAKU_KATAKANA_LAST_CHAR) {
			return ZENKAKU_KATAKANA[c - HANKAKU_KATAKANA_FIRST_CHAR];
		} else {
			return c;
		}
	}

	public static char mergeChar(char c1, char c2) {
		if (c2 == 'ﾞ') {
			if ("ｶｷｸｹｺｻｼｽｾｿﾀﾁﾂﾃﾄﾊﾋﾌﾍﾎ".indexOf(c1) > 0) {
				switch (c1) {
				case 'ｶ':
					return 'ガ';
				case 'ｷ':
					return 'ギ';
				case 'ｸ':
					return 'グ';
				case 'ｹ':
					return 'ゲ';
				case 'ｺ':
					return 'ゴ';
				case 'ｻ':
					return 'ザ';
				case 'ｼ':
					return 'ジ';
				case 'ｽ':
					return 'ズ';
				case 'ｾ':
					return 'ゼ';
				case 'ｿ':
					return 'ゾ';
				case 'ﾀ':
					return 'ダ';
				case 'ﾁ':
					return 'ヂ';
				case 'ﾂ':
					return 'ヅ';
				case 'ﾃ':
					return 'デ';
				case 'ﾄ':
					return 'ド';
				case 'ﾊ':
					return 'バ';
				case 'ﾋ':
					return 'ビ';
				case 'ﾌ':
					return 'ブ';
				case 'ﾍ':
					return 'ベ';
				case 'ﾎ':
					return 'ボ';
				}
			}
		} else if (c2 == 'ﾟ') {
			if ("ﾊﾋﾌﾍﾎ".indexOf(c1) > 0) {
				switch (c1) {
				case 'ﾊ':
					return 'パ';
				case 'ﾋ':
					return 'ピ';
				case 'ﾌ':
					return 'プ';
				case 'ﾍ':
					return 'ペ';
				case 'ﾎ':
					return 'ポ';
				}
			}
		}
		return c1;
	}

	public static String hankakuKatakanaToZenkakuKatakana(String s) {
		if (s.length() == 0) {
			return s;
		} else if (s.length() == 1) {
			return hankakuKatakanaToZenkakuKatakana(s.charAt(0)) + "";
		} else {
			StringBuffer sb = new StringBuffer(s);
			int i = 0;
			for (i = 0; i < sb.length() - 1; i++) {
				char originalChar1 = sb.charAt(i);
				char originalChar2 = sb.charAt(i + 1);
				char margedChar = mergeChar(originalChar1, originalChar2);
				if (margedChar != originalChar1) {
					sb.setCharAt(i, margedChar);
					sb.deleteCharAt(i + 1);
				} else {
					char convertedChar = hankakuKatakanaToZenkakuKatakana(originalChar1);
					if (convertedChar != originalChar1) {
						sb.setCharAt(i, convertedChar);
					}
				}
			}
			if (i < sb.length()) {
				char originalChar1 = sb.charAt(i);
				char convertedChar = hankakuKatakanaToZenkakuKatakana(originalChar1);
				if (convertedChar != originalChar1) {
					sb.setCharAt(i, convertedChar);
				}
			}
			return sb.toString();
		}
	}

	public static String zenkakuHiraganaToZenkakuKatakana(String s) {
		StringBuffer sb = new StringBuffer(s);
		for (int i = 0; i < sb.length(); i++) {
			char c = sb.charAt(i);
			if (c >= 'ぁ' && c <= 'ん') {
				sb.setCharAt(i, (char) (c - 'ぁ' + 'ァ'));
			}
		}
		return sb.toString();
	}

	public static String convertToKatakana(String s) {
		s = hankakuKatakanaToZenkakuKatakana(s);
		s = zenkakuHiraganaToZenkakuKatakana(s);
		return s;
	}
}

File: app/src/main/java/com/github/yuukis/businessmap/util/StringJUtils.kt
package com.github.yuukis.businessmap.util

object StringJUtils {

    private val HANKAKU_KATAKANA = charArrayOf(
        '｡', '｢', '｣', '､', '･',
        'ｦ', 'ｧ', 'ｨ', 'ｩ', 'ｪ', 'ｫ', 'ｬ', 'ｭ', 'ｮ', 'ｯ', 'ｰ', 'ｱ', 'ｲ',
        'ｳ', 'ｴ', 'ｵ', 'ｶ', 'ｷ', 'ｸ', 'ｹ', 'ｺ', 'ｻ', 'ｼ', 'ｽ', 'ｾ', 'ｿ',
        'ﾀ', 'ﾁ', 'ﾂ', 'ﾃ', 'ﾄ', 'ﾅ', 'ﾆ', 'ﾇ', 'ﾈ', 'ﾉ', 'ﾊ', 'ﾋ', 'ﾌ',
        'ﾍ', 'ﾎ', 'ﾏ', 'ﾐ', 'ﾑ', 'ﾒ', 'ﾓ', 'ﾔ', 'ﾕ', 'ﾖ', 'ﾗ', 'ﾘ', 'ﾙ',
        'ﾚ', 'ﾛ', 'ﾜ', 'ﾝ', 'ﾞ', 'ﾟ'
    )

    private val ZENKAKU_KATAKANA = charArrayOf(
        '。', '「', '」', '、', '・',
        'ヲ', 'ァ', 'ィ', 'ゥ', 'ェ', 'ォ', 'ャ', 'ュ', 'ョ', 'ッ', 'ー', 'ア', 'イ',
        'ウ', 'エ', 'オ', 'カ', 'キ', 'ク', 'ケ', 'コ', 'サ', 'シ', 'ス', 'セ', 'ソ',
        'タ', 'チ', 'ツ', 'テ', 'ト', 'ナ', 'ニ', 'ヌ', 'ネ', 'ノ', 'ハ', 'ヒ', 'フ',
        'ヘ', 'ホ', 'マ', 'ミ', 'ム', 'メ', 'モ', 'ヤ', 'ユ', 'ヨ', 'ラ', 'リ', 'ル',
        'レ', 'ロ', 'ワ', 'ン', '゛', '゜'
    )

    private val HANKAKU_KATAKANA_FIRST_CHAR = HANKAKU_KATAKANA[0]

    private val HANKAKU_KATAKANA_LAST_CHAR = HANKAKU_KATAKANA[HANKAKU_KATAKANA.size - 1]

    /**
     * 半角カタカナから全角カタカナへ変換します。
     *
     * @param c 変換前の文字
     * @return 変換後の文字
     */
    @JvmStatic
    fun hankakuKatakanaToZenkakuKatakana(c: Char): Char {
        return if (c in HANKAKU_KATAKANA_FIRST_CHAR..HANKAKU_KATAKANA_LAST_CHAR) {
            ZENKAKU_KATAKANA[c - HANKAKU_KATAKANA_FIRST_CHAR]
        } else {
            c
        }
    }

    @JvmStatic
    fun mergeChar(c1: Char, c2: Char): Char {
        if (c2 == 'ﾞ') {
            if ("ｶｷｸｹｺｻｼｽｾｿﾀﾁﾂﾃﾄﾊﾋﾌﾍﾎ".indexOf(c1) >= 0) {
                return when (c1) {
                    'ｶ' -> 'ガ'
                    'ｷ' -> 'ギ'
                    'ｸ' -> 'グ'
                    'ｹ' -> 'ゲ'
                    'ｺ' -> 'ゴ'
                    'ｻ' -> 'ザ'
                    'ｼ' -> 'ジ'
                    'ｽ' -> 'ズ'
                    'ｾ' -> 'ゼ'
                    'ｿ' -> 'ゾ'
                    'ﾀ' -> 'ダ'
                    'ﾁ' -> 'ヂ'
                    'ﾂ' -> 'ヅ'
                    'ﾃ' -> 'デ'
                    'ﾄ' -> 'ド'
                    'ﾊ' -> 'バ'
                    'ﾋ' -> 'ビ'
                    'ﾌ' -> 'ブ'
                    'ﾍ' -> 'ベ'
                    'ﾎ' -> 'ボ'
                    else -> c1
                }
            }
        } else if (c2 == 'ﾟ') {
            if ("ﾊﾋﾌﾍﾎ".indexOf(c1) >= 0) {
                return when (c1) {
                    'ﾊ' -> 'パ'
                    'ﾋ' -> 'ピ'
                    'ﾌ' -> 'プ'
                    'ﾍ' -> 'ペ'
                    'ﾎ' -> 'ポ'
                    else -> c1
                }
            }
        }
        return c1
    }

    @JvmStatic
    fun hankakuKatakanaToZenkakuKatakana(s: String): String {
        if (s.isEmpty()) {
            return s
        } else if (s.length == 1) {
            return hankakuKatakanaToZenkakuKatakana(s[0]).toString()
        } else {
            val sb = StringBuilder(s)
            var i = 0
            while (i < sb.length - 1) {
                val originalChar1 = sb[i]
                val originalChar2 = sb[i + 1]
                val mergedChar = mergeChar(originalChar1, originalChar2)
                if (mergedChar != originalChar1) {
                    sb.setCharAt(i, mergedChar)
                    sb.deleteCharAt(i + 1)
                } else {
                    val convertedChar = hankakuKatakanaToZenkakuKatakana(originalChar1)
                    if (convertedChar != originalChar1) {
                        sb.setCharAt(i, convertedChar)
                    }
                }
                i++
            }
            if (i < sb.length) {
                val originalChar1 = sb[i]
                val convertedChar = hankakuKatakanaToZenkakuKatakana(originalChar1)
                if (convertedChar != originalChar1) {
                    sb.setCharAt(i, convertedChar)
                }
            }
            return sb.toString()
        }
    }

    @JvmStatic
    fun zenkakuHiraganaToZenkakuKatakana(s: String): String {
        val sb = StringBuilder(s)
        for (i in sb.indices) {
            val c = sb[i]
            if (c in 'ぁ'..'ん') {
                sb.setCharAt(i, (c - 'ぁ' + 'ァ'.code).toChar())
            }
        }
        return sb.toString()
    }

    @JvmStatic
    fun convertToKatakana(s: String): String {
        var result = hankakuKatakanaToZenkakuKatakana(s)
        result = zenkakuHiraganaToZenkakuKatakana(result)
        return result
    }
}
